### PR TITLE
OpenSSL FIPS provider support

### PR DESCRIPTION
**Pull request by jbdelcuv**
Add support to the SGX-SSL library to have crypto algorithms run in the FIPS provider embedded in enclave images.

Instructions for building/testing, assuming you have an updated SGX SDK/PSW toolchain available:
```
$ cd Linux
$ make all FIPS=1
$ [sudo] make install
$ make test FIPS=1
$ make fips_test FIPS=1
```

Instructions for cleaning up:
```
$ sudo make uninstall
$ make clean
```

### CODE REVIEW

**Comment by fchinchilla** [comment]
Looks good to me

File: .github/workflows/c-cpp.yml [hunk 1]
```diff
@@ -2,9 +2,9 @@ name: C/C++ CI
 
 on:
   push:
-    branches: [ "main" ]
+    branches: [ "support_openssl_3.1*" ]
```
Comment (lzha101) [inline .github/workflows/c-cpp.yml:5]: Why change the branch from main to support_openssl_3.1*?

File: .github/workflows/codeql.yml [hunk 2]
```diff
@@ -13,11 +13,11 @@ name: "CodeQL"
 
 on:
   push:
-    branches: [ "main" ]
+    branches: [ "support_openssl_3.1*" ]
```
Comment (lzha101) [inline .github/workflows/codeql.yml:16]: Do we need to revert the changes from Line 16 to Line20?

File: .github/workflows/c-cpp.yml (hunk 1 above)
Comment (jbdelcuv) [reply .github/workflows/c-cpp.yml:5]: Yes, that makes sense.
I'll update the PR.

File: .github/workflows/codeql.yml (hunk 2 above)
Comment (jbdelcuv) [reply .github/workflows/codeql.yml:16]: Yes, that makes sense.
I'll update the PR.

File: .github/workflows/c-cpp.yml [hunk 3]
```diff
@@ -25,7 +25,7 @@ jobs:
     - name: install SDK
       run: chmod +x *.bin; echo yes | ./sgx_linux_x64_sdk_2.*.bin
     - name: download OpenSSL code
-      run: wget https://www.openssl.org/source/openssl-3.0.14.tar.gz --directory-prefix=openssl_source/
+      run: wget https://www.openssl.org/source/openssl-3.1.5.tar.gz --directory-prefix=openssl_source/
```
Comment (jbdelcuv) [inline .github/workflows/c-cpp.yml:28]: I'll also update this to version 3.1.6, which is what we intend to support.

File: Linux/sgx/fips_test/Makefile [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+#
+include ../buildenv.mk
+
+all:
```
Comment (lzha101) [inline Linux/sgx/fips_test/Makefile:33]: What's the difference between this fips_test and test_app? Looks the test_app also includes fips functionality tests.

Comment (jbdelcuv) [reply Linux/sgx/fips_test/Makefile:33]: The fips_test is based on the test_app but is a smaller program to help focusing on the APIs that an enclave has to call in order to initialize the OSSL FIPS provider.

File: Linux/sgx/test_app/enclave/TestEnclave.cpp [hunk 5]
```diff
@@ -295,130 +293,188 @@ void t_sgxssl_call_apis()
     printf("Start tests\n");
     
     SGXSSLSetPrintToStdoutStderrCB(vprintf_cb);
+    OSSL_PROVIDER *prov;
+#ifndef SGXSSL_FIPS
+    prov = OSSL_PROVIDER_load(NULL, "default");
+#else
+    void *entry = get_ossl_fips_sym("OSSL_provider_init");
```
Comment (lzha101) [inline Linux/sgx/test_app/enclave/TestEnclave.cpp:300]: The API is updated to 'sgx_get_ossl_fips_sym'. Please update this and the one in fips_test.

Comment (jbdelcuv) [reply Linux/sgx/test_app/enclave/TestEnclave.cpp]: Pushed one commit to rename the API.

File: Linux/sgx/fips_test/sgx_t.mk [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	@echo "There is no enclave test key <enclave_private.pem>."
+	@echo "The project will generate a key <enclave_private.pem> for testing."
+	@openssl genrsa -out $(Enclave_Test_Key) -3 3072
+endif
+	@echo "SIGN =>  $@"
+	$(SGX_ENCLAVE_SIGNER) sign -key $(Enclave_Test_Key) -enclave enclave.so -out $@ -config $(ENCLAVE_DIR)/enclave.config.xml
+
+install_conf:
+	@echo "*** Installing OpenSSL configuration"
+	@echo "install $(OPENSSLCONF) -> $(SGX_SDK)/$(LIBDIR)/$(OPENSSLCONF)"
```
Comment (lzha101) [inline Linux/sgx/fips_test/sgx_t.mk:181]: Line 181 looks incorrect and could be removed.

File: Linux/sgx/fips_test/sgx_t.mk [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	@cp -f $(OPENSSLCONF).tmpl $(OPENSSLCONF)
+	echo ".include $(SGX_SDK)/$(LIBDIR)/$(FIPSMODULECONF)" >> $(OPENSSLCONF)
+
+clean:
+	@rm -f enclave.* $(ENCLAVE_DIR)/enclave_t.* $(Enclave_Cpp_Objects) $(Enclave_C_Objects) $(Enclave_Test_Key)
+	@rm -f fips.so openssl.cnf
```
Comment (lzha101) [inline Linux/sgx/fips_test/sgx_t.mk:187]: No fips.so is copied, right?

Comment (jbdelcuv) [reply Linux/sgx/fips_test/sgx_t.mk:187]: Not in this makefile.
The fips.so is copied in the open_source/Makefile lines 73--77.

File: Linux/sgx/fips_test/sgx_t.mk (hunk 6 above)
Comment (jbdelcuv) [reply Linux/sgx/fips_test/sgx_t.mk:181]: It's a simple echo command to indicate the user what the Makefile is going to do.
The OpenSSL library does something similar.
Why do you think is incorrect?

Comment (lzha101) [reply Linux/sgx/fips_test/sgx_t.mk:181]: The message is saying to install the openssl.cnf to $(SGX_SDK)/$(LIBDIR). This is not current behavior, right? Current behavior is to generate an openssl.cnf in the test directory.

Comment (jbdelcuv) [reply Linux/sgx/fips_test/sgx_t.mk:181]: You're right, too much copy & paste, the second part of the message "-> $(SGX_SDK)/$(LIBDIR)/$(OPENSSLCONF)" should be removed.
